pfs_middleware/pfs_middleware/middleware.py

```diff
@@ -438,30 +439,111 @@ def extract_container_metadata_from_headers(req):
-    if is_dir:
-        return EMPTY_OBJECT_ETAG
-    if container_listing and LISTING_ETAG_OVERRIDE_HEADER in obj_metadata:
+def mung_etags(obj_metadata, etag, num_writes):
+    '''Mung the ETag headers that will be stored with an object.  The
+    goal is to preserve ETag metadata passed down by other filters but
+    to do so in such a way that it will be invalidated if there is a
+    write to or truncate of the object via the ProxyFS file API.
+
+    The mechanism is to prepend a counter to the ETag header values
+    that is incremented each time the object is modified is modified.
+    When the object is read, if the value for the counter has changed,
+    the ETag is assumed to be invalid.  The counter is typically the
+    number of writes to the object.
+
+    etag is either None or the value that should be returned as the
+    ETag for the object (in the absence of other considerations).
+
+    This assumes that all headers have been converted to "titlecase",
+    which means, among other things, that "ETag" will show up as
+    "Etag".
+
+    This ignores SLO headers because it assumes they have already been
+    stripped.
+    '''
+    if LISTING_ETAG_OVERRIDE_HEADER in obj_metadata:
+        obj_metadata[LISTING_ETAG_OVERRIDE_HEADER] = "%d:%s" % (
+            num_writes, obj_metadata[LISTING_ETAG_OVERRIDE_HEADER])
+
+    if S3API_ETAG_HEADER in obj_metadata:
+        obj_metadata[S3API_ETAG_HEADER] = "%d:%s" % (
+            num_writes, obj_metadata[S3API_ETAG_HEADER])
+
+    if etag is not None:
+        obj_metadata[ORIGINAL_MD5_HEADER] = "%d:%s" % (num_writes, etag)
+
+    return
+
+
+def unmung_etags_and_return_best(obj_metadata, account_name, inum, num_writes,
+                                 is_dir=False, container_listing=False):
+    '''Unmung the ETag headers associated with an object to return
+    them to the state they were in when passed to pfs_middleware or
+    delete them if the object has changed.
+
+    Return the ETag that is most likely to be correct for the query,
+    but leave other valid ETags values in the metadata, in case a
+    higher layer filter wants to use them to override the value
+    returned here.
+
+    If the ETags in the metadata are invalid, construct and return a
+    new ProxyFS ETag based on the account name, inode number, and
+    number of writes.
+    '''
+
+    # purge headers made invalid by subsequent writes/truncates from
+    # the returned headers
+    if LISTING_ETAG_OVERRIDE_HEADER in obj_metadata:
         val = obj_metadata[LISTING_ETAG_OVERRIDE_HEADER]
         try:
             stored_num_writes, rest = val.split(':', 1)
             if int(stored_num_writes) == num_writes:
-                return rest
+                obj_metadata[LISTING_ETAG_OVERRIDE_HEADER] = rest
+            else:
+                del obj_metadata[LISTING_ETAG_OVERRIDE_HEADER]
         except ValueError:
+            del obj_metadata[LISTING_ETAG_OVERRIDE_HEADER]
             pass
```

Comment:
good point!  i'm not that good at python ...

Reply:
sorry, i didn't address all of your comment!

anyway, if the conditional ETag header is no longer valid because somebody/something has written to the file, why do we want to keep it?  it just uses up disk space and network bandwidth.

is there a reason to keep `LISTING_ETAG_OVERRIDE_HEADER` once it is no longer germane?

Reply:
Are we talking about keeping it persisted in the metadata stream in proxyfs? I'd assumed because we didn't want to have to go digging through the metadata to figure out what's still valid and what isn't on every write.

Or are we talking about in the request path? We definitely don't need to keep it around then; the `.pop()` will do just that:

```
>>> x={'foo': 'bar'}
>>> x
{'foo': 'bar'}
>>> x.pop('foo')
'bar'
>>> x
{}
```
```

Reply:
Ahhh.  yes, the .pop() is better.